### DISCUSSION TITLE
Multithreading architecture [13712]

### DISCUSSION THREAD

**Discussion by JerrySciangula** [eProsima/Fast-DDS#2456] [Q&A]
Hi all,

I'm a PhD student at Scuola Superiore Sant'Anna in Pisa. I'm currently working on FastDDS.

Checking the code and the specification, I figured out the basic multithreading architecture, but I still have some doubts:

1) how many Reception Threads are spawn by each Publisher or Subscriber? (I know that the main thread can spawn a single Resource Thread and (in case of a Writer) a single Asynchronous thread for the sending of the messages asynchronously). I tried the deadlineQoSExample and I found:

-           The Main Thread
-           The FlowController Thread
-           The ResourceEvent Thread
-           A SharedMemWatchdog Thread
-           2 SharedMemChannelResource Threads
-           3 UDPChannelResource Threads

Why 3 Threads for UDP channel communication are used? Why 2 for the Shared Memory Channel?
 
2) Are all the threads created at init time?

3) Is there any file or documentation where the relationship between threads are highlighted?

Thank you in advance and for your help.

**Comment by JLBuenoLopez** [comment]
Maybe this answer may be of help: https://github.com/eProsima/Fast-DDS/issues/1769#issuecomment-781135230
